### PR TITLE
crypt: add KDF options to API

### PR DESCRIPTION
**Pull request by chrisccoulson**
This is based on https://github.com/snapcore/secboot/pull/163 with some of the comments addressed.

### CODE REVIEW

File: crypt.go [hunk 1]
```diff
@@ -692,6 +735,16 @@ func InitializeLUKS2Container(devicePath, label string, key []byte, options *Ini
 	if len(key) < 32 {
 		return fmt.Errorf("expected a key length of at least 256-bits (got %d)", len(key)*8)
 	}
+
+	// Simplify things a bit
+	defaultKdfOptions := &KDFOptions{TargetDuration: 100 * time.Millisecond}
```
Comment (mvo5) [inline crypt.go:740]: Thanks for this too

**Review by mvo5 — APPROVED** [review]
Thank you

File: crypt.go [hunk 2]
```diff
@@ -628,6 +629,50 @@ func setLUKS2KeyslotPreferred(devicePath string, slot int) error {
 	return nil
 }
 
+// AddRecoveryKeyToLUKS2ContainerOptions carries options for recovery key to LUKS2
+// containers.
+type KDFOptions struct {
+	// PbkdfMmemory set the memory cost for PBKDF (for Argon2i/id the number
```
Comment (pedronis) [inline crypt.go:635]: Pkbdf should be dropped I suppose, and also PB from the PBKDFs ...

File: crypt.go [hunk 3]
```diff
@@ -628,6 +629,50 @@ func setLUKS2KeyslotPreferred(devicePath string, slot int) error {
 	return nil
 }
 
+// AddRecoveryKeyToLUKS2ContainerOptions carries options for recovery key to LUKS2
+// containers.
```
Comment (pedronis) [inline crypt.go:633]: this needs different content

**Review by pedronis — APPROVED** [review]
lgtm, the comments need some tweaks

File: crypt.go [hunk 4]
```diff
@@ -628,6 +629,57 @@ func setLUKS2KeyslotPreferred(devicePath string, slot int) error {
...
+
+	if options.MemoryKiB != 0 {
+		args = append(args,
+			"--pbkdf-memory", strconv.Itoa(options.MemoryKiB))
+	}
+
+	switch {
+	case options.ForceIterations != 0:
+		// Disable benchmarking by forcing the time cost.
+		args = append(args,
```
Comment (pedronis) [inline crypt.go:668]: I suppose this path could use a new test

**Review by pedronis — APPROVED** [review]
thanks

File: crypt.go [hunk 5]
```diff
@@ -795,10 +862,11 @@ func addKeyToLUKS2Container(devicePath string, existingKey, key []byte, extraOpt
 // key argument.
 //
 // The recovery key is provided via the recoveryKey argument and must be a cryptographically secure 16-byte number.
-func AddRecoveryKeyToLUKS2Container(devicePath string, key []byte, recoveryKey RecoveryKey) error {
-	return addKeyToLUKS2Container(devicePath, key, recoveryKey[:], []string{
-		// use argon2i as the KDF with an increased cost
-		"--pbkdf", "argon2i", "--iter-time", "5000"})
+func AddRecoveryKeyToLUKS2Container(devicePath string, key []byte, recoveryKey RecoveryKey, options *KDFOptions) error {
+	if options == nil {
+		options = &KDFOptions{TargetDuration: 5 * time.Second}
```
Comment (mvo5) [inline crypt.go:867]: Didn't we agree to move this to 2s instead?

**Review by mvo5 — COMMENTED** [review]
One more question

**Review by mvo5 — APPROVED** [review]
\o/ thanks again

**Pull request merged**